### PR TITLE
Add 'conda-remove-defaults' setting and support 'nodefaults' as a keyword channel

### PR DESCRIPTION
**Pull request by jaimergp**
Supersedes https://github.com/conda-incubator/setup-miniconda/pull/364

### CODE REVIEW

**Comment by jaimergp** [comment]
Errors in the mamba examples are unrelated and caused by the mamba 2.0 release not having `mamba.bat` under `condabin/` anymore. Adding it at https://github.com/conda-forge/mamba-feedstock/pull/251

**Comment by jakirkham** [comment]
Thanks Jaime! 🙏

Am actually wondering whether we need a setting for this or if we should just treat an explicit channel as just that what a user intended

Think it is easier and clearer for users to add `defaults` if they want it in the channel list

**Comment by jaimergp** [comment]
Yes, in the long term, that's where I want to get at. My current concern if that if we change that behaviour overnight without warning first, then we might break a few CI pipelines that were inadvertently relying on defaults as a channel (because [strict priority is not default either](https://github.com/conda/conda/issues/14274)). 

After all this action is relied on by...  [16k+ workflows](https://github.com/search?q=%22uses%3A+conda-incubator%2Fsetup-miniconda%22+path%3A.github%2Fworkflows%2F&ref=opensearch&type=code). Of those, [~6k set `channels`](https://github.com/search?q=%22uses%3A+conda-incubator%2Fsetup-miniconda%22+path%3A.github%2Fworkflows%2F+%22channels%3A%22&ref=opensearch&type=code). Of those, [~3.7k workflows specify `defaults` explicitly](https://github.com/search?q=%22uses%3A+conda-incubator%2Fsetup-miniconda%22+path%3A.github%2Fworkflows%2F+%2Fchannels%3A.*defaults.*%2F&ref=opensearch&type=code).

**Comment by jaimergp** [comment]
Fix for example 6 available at #368. Once that's merged we can continue here.

**Review by dbast — APPROVED** [review]

File: action.yml [hunk 1]
```diff
@@ -193,6 +193,12 @@ inputs:
       for more information."
     required: false
     default: ""
+  no-implicit-channels:
```
Comment (jezdez) [inline action.yml:196]:
```suggestion
  conda-remove-defaults:
```

File: README.md [hunk 2]
```diff
@@ -598,6 +603,35 @@ jobs:
           python -c "import platform; assert platform.machine() == 'arm64', platform.machine()"
 ```
 
+### Example 14: No implicit channels
```
Comment (jezdez) [inline README.md:606]:
```suggestion
### Example 14: Remove `defaults` channel
```

File: README.md [hunk 3]
```diff
@@ -598,6 +603,35 @@ jobs:
...
+        os: ["ubuntu-latest"]
+    steps:
+      - uses: actions/checkout@v4
+      - uses: ./
+        id: setup-miniconda
+        continue-on-error: true
+        with:
+          miniforge-version: latest
+          channels: conda-forge
+          no-implicit-channels: "true"
```
Comment (jezdez) [inline README.md:628]:
```suggestion
          conda-remove-defaults: "true"
```

File: .github/workflows/example-14.yml [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    timeout-minutes: 30
+    steps:
+      - uses: actions/checkout@v4
+      - uses: ./
+        id: setup-miniconda
+        continue-on-error: true
+        with:
+          channels: conda-forge
+          python-version: "3.12"
+          no-implicit-channels: ${{ matrix.no-implicit-channels }}
```
Comment (jezdez) [inline .github/workflows/example-14.yml:43]:
```suggestion
          conda-remove-defaults: ${{ matrix.conda-remove-defaults }}
```

File: .github/workflows/example-14.yml [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    if:
+      (github.event_name == 'schedule' && github.repository ==
+      'conda-incubator/setup-miniconda') || (github.event_name != 'schedule')
+    name: Ex14 (${{ matrix.os }}, ${{ matrix.no-implicit-channels }})
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
+        no-implicit-channels: ["true", "false"]
```
Comment (jezdez) [inline .github/workflows/example-14.yml:33]:
```suggestion
        conda-remove-defaults: ["true", "false"]
```

File: .github/workflows/example-14.yml [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    ${{ github.workflow }}-${{ github.event.pull_request.number || github.sha }}
+  cancel-in-progress: true
+
+jobs:
+  example-14:
+    # prevent cronjobs from running on forks
+    if:
+      (github.event_name == 'schedule' && github.repository ==
+      'conda-incubator/setup-miniconda') || (github.event_name != 'schedule')
+    name: Ex14 (${{ matrix.os }}, ${{ matrix.no-implicit-channels }})
```
Comment (jezdez) [inline .github/workflows/example-14.yml:27]:
```suggestion
    name: Ex14 (${{ matrix.os }}, ${{ matrix.conda-remove-defaults }})
```

File: .github/workflows/example-14.yml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+        continue-on-error: true
+        with:
+          channels: conda-forge
+          python-version: "3.12"
+          no-implicit-channels: ${{ matrix.no-implicit-channels }}
+      - shell: bash -el {0}
+        run: |
+          channels=$(conda config --show channels --json | jq -r '.channels | join(",")')
+          echo $channels
+          if [[ "${{ matrix.no-implicit-channels }}" == "true" ]]; then
```
Comment (jezdez) [inline .github/workflows/example-14.yml:48]:
```suggestion
          if [[ "${{ matrix.conda-remove-defaults }}" == "true" ]]; then
```

File: README.md [hunk 8]
```diff
@@ -59,6 +59,7 @@ possibility of automatically activating the `test` environment on all shells.
 | [Caching packages](#caching-packages)                              | [![Caching Example Status][caching-badge]][caching]             |
 | [Caching environments](#caching-environments)                      | [![Caching Env Example Status][caching-env-badge]][caching-env] |
 | [Apple Silicon](#example-13-apple-silicon)                         | [![Apple Silicon][ex13-badge]][ex13]                            |
+| [No implicit channels](#example-14-no-implicit-channels)           | [![No implicit channels][ex14-badge]][ex14]                     |
```
Comment (jezdez) [inline README.md:62]:
```suggestion
| [Remove defaults](#example-14-conda-remove-defaults)           | [![Remove defaults][ex14-badge]][ex14]                     |
```

File: src/input.ts [hunk 9]
```diff
@@ -92,6 +92,7 @@ export async function parseInputs(): Promise<types.IActionInputs> {
     minicondaVersion: core.getInput("miniconda-version"),
     miniforgeVariant: core.getInput("miniforge-variant"),
     miniforgeVersion: core.getInput("miniforge-version"),
+    noImplicitChannels: core.getInput("no-implicit-channels"),
```
Comment (jezdez) [inline src/input.ts:95]:
```suggestion
    condaRemoveDefaults: core.getInput("conda-remove-defaults"),
```

File: src/types.ts [hunk 10]
```diff
@@ -84,6 +84,7 @@ export interface IActionInputs {
   readonly minicondaVersion: string;
   readonly miniforgeVariant: string;
   readonly miniforgeVersion: string;
+  readonly noImplicitChannels: string;
```
Comment (jezdez) [inline src/types.ts:87]:
```suggestion
  readonly condaRemoveDefaults: string;
```

File: src/conda.ts [hunk 11]
```diff
@@ -166,6 +166,20 @@ export async function applyCondaConfiguration(
     await condaCommand(["config", "--add", "channels", channel], options);
   }
 
+  if (inputs.noImplicitChannels === "true" && !channels.includes("defaults")) {
```
Comment (jezdez) [inline src/conda.ts:169]:
```suggestion
  if (inputs.condaRemoveDefaults === "true" && !channels.includes("defaults")) {
```

File: dist/setup/index.js [hunk 12]
```diff
@@ -47355,6 +47355,15 @@ function applyCondaConfiguration(inputs, options) {
             core.info(`Adding channel '${channel}'`);
             yield condaCommand(["config", "--add", "channels", channel], options);
         }
+        if (inputs.noImplicitChannels === "true" && !channels.includes("defaults")) {
```
Comment (jezdez) [inline dist/setup/index.js:47358]:
```suggestion
        if (inputs.condaRemoveDefaults === "true" && !channels.includes("defaults")) {
```

File: dist/setup/index.js [hunk 13]
```diff
@@ -48188,6 +48197,7 @@ function parseInputs() {
             minicondaVersion: core.getInput("miniconda-version"),
             miniforgeVariant: core.getInput("miniforge-variant"),
             miniforgeVersion: core.getInput("miniforge-version"),
+            noImplicitChannels: core.getInput("no-implicit-channels"),
```
Comment (jezdez) [inline dist/setup/index.js:48200]:
```suggestion
            condaRemoveDefaults: core.getInput("conda-remove-defaults"),
```

File: action.yml [hunk 14]
```diff
@@ -193,6 +193,12 @@ inputs:
...
+    description:
+      "Postprocess channels list to remove channels that were not added
+      explicitly in the 'channels' setting. Will default to 'true' in 2025.03."
```
Comment (jezdez) [inline action.yml:199]:
```suggestion
      "Postprocess channels list to remove 'defaults' if it was added
      implicitely to the 'channels' setting. Will default to 'true' to
      aligned with related change in conda 25.3.0."
```

**Review by jezdez — COMMENTED** [review]
I'd suggest to be.. more explicit with the option name, since all it does right now is removing "defaults", nothing else. Also prefix the option with `conda` since this only applies to conda, not mamba.

File: .github/workflows/example-14.yml [hunk 15]
```diff
@@ -0,0 +1,54 @@
+name: "Example 14: Remove implicitly added channels"
```
Comment (jaimergp) [inline .github/workflows/example-14.yml:1]:
```suggestion
name: "Example 14: Remove defaults channel"
```

File: .github/workflows/example-14.yml [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+    if:
+      (github.event_name == 'schedule' && github.repository ==
+      'conda-incubator/setup-miniconda') || (github.event_name != 'schedule')
+    name: Ex14 (${{ matrix.os }}, ${{ matrix.conda-remove-defaults }})
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os: ["ubuntu-latest", "macos-latest", "windows-latest"]
+        conda-remove-defaults: ["true", "false"]
```
Comment (jakirkham) [inline .github/workflows/example-14.yml:33]: How about something like

```suggestion
        explicitly-listed-channels-only: ["true", "false"]
```

Admittedly verby, but I think we can write this in a way that gets closer to the long-term intended behavior

Comment (jaimergp) [reply .github/workflows/example-14.yml:35]: See https://github.com/conda-incubator/setup-miniconda/pull/367#pullrequestreview-2347356646. I don't particularly care about the name itself, given it'll be probably be obsolete by March next year.

**Comment by jakirkham** [comment]
> Yes, in the long term, that's where I want to get at. My current concern if that if we change that behaviour overnight without warning first, then we might break a few CI pipelines that were inadvertently relying on defaults as a channel (because [strict priority is not default either](https://github.com/conda/conda/issues/14274)).
> 
> After all this action is relied on by... [16k+ workflows](https://github.com/search?q=%22uses%3A+conda-incubator%2Fsetup-miniconda%22+path%3A.github%2Fworkflows%2F&ref=opensearch&type=code). Of those, [~6k set `channels`](https://github.com/search?q=%22uses%3A+conda-incubator%2Fsetup-miniconda%22+path%3A.github%2Fworkflows%2F+%22channels%3A%22&ref=opensearch&type=code). Of those, [~3.7k workflows specify `defaults` explicitly](https://github.com/search?q=%22uses%3A+conda-incubator%2Fsetup-miniconda%22+path%3A.github%2Fworkflows%2F+%2Fchannels%3A.*defaults.*%2F&ref=opensearch&type=code).

AIUI the path proposed here is...

1. Add a new flag and document
2. Recommend users start setting (maybe with a warning)
3. Change the default in a major version
4. Deprecate the flag
5. Remove the flag (in the next major version)

It is certainly a valid way to do it. Am just concerned by the length of time that will take and the effort it puts us and users. Can we cutdown this process a bit?

Admittedly there is the other extreme, change behavior in a major version with documentation. This is not an unheard of approach though we need not do that either

Think there is a reasonable middle path. We already are deprecating `nodefaults`. Why not leverage it for one last release then do a major version bump that drops it everywhere?

If not that, can we thin this process a bit? How can we get this to one deprecation cycle?

**Comment by jaimergp** [comment]
In a way we are tied to the conda deprecation cycle, where this behavior won't be default til March 2025. We might need to keep the flag around for users of older versions.

> Think there is a reasonable middle path. We already are deprecating nodefaults. Why not leverage it for one last release then do a major version bump that drops it everywhere?

This looks reasonable to me, but the major version bump should happen in March to align with conda. Is that too long?

**Comment by jaimergp** [comment]
This should be ready now:

- [Example warning for nodefaults](https://github.com/conda-incubator/setup-miniconda/actions/runs/11178230690/job/31075405441?pr=367#step:3:31)
- [Example info for conda-remove-defaults](https://github.com/conda-incubator/setup-miniconda/actions/runs/11178230690/job/31075404577?pr=367#step:3:33)
- [Example warning for implicit defaults](https://github.com/conda-incubator/setup-miniconda/actions/runs/11178230690/job/31075406250?pr=367#step:3:33)

**Comment by jezdez** [comment]
I strongly disagree that `nodefaults` should be elevated to a first class feature in the action. I'm lukewarm even acknowledging it with a warning, but I guess that's fine. My main concern is that this is feature creep for a very short timeframe.

**Comment by jaimergp** [comment]
Yea, I don't like it either, but there are [~200 workflows out there using it](https://github.com/search?q=%22uses%3A+conda-incubator%2Fsetup-miniconda%22+path%3A.github%2Fworkflows%2F+%2Fchannels%3A.*nodefaults.*%2F&ref=opensearch&type=code). We need to warn them about the new option, enable the behavior, and then turn it into an error. 

It's not even documented in the README, while `conda-remove-defaults` is. I think this is the best compromise we can afford for now.

File: .github/workflows/example-14.yml (hunk 16 above)
Comment (jezdez) [reply .github/workflows/example-14.yml:35]: I think the long-term future for this feature is only to migrate conda users to the same pattern as mamba and pixi follow, setting a default channel in the distribution/bootstrapper layer and not hard-coding it in the codebase. That's why I think the name of the parameter for setup-miniconda is okay to focus on conda only, and not required to be named generically.

**Review by jezdez — APPROVED** [review]

**Pull request merged**